### PR TITLE
Implement masking during sampling to improve blending

### PR DESCRIPTION
**Pull request by anon-hlhl**
- Almost inpainting-like behaviour
 - Works best with small erased areas using "Regenerate masked area"
 - Denoiser strength to 1 results to completely in-paint with no previous image info
 - Uncrop kind of sometimes works, but not really (use with Denoiser strength 1)
 - Works with k-diff samplers & DDIM, but...
 - DDIM changes depend on a PR in stable-diffusion core (https://github.com/hlky/stable-diffusion/pull/127)

Still a victim of the weird gradio masking editor bugs

Silly demo

### CODE REVIEW

File: frontend/ui_functions.py [hunk 1]
```diff
@@ -8,8 +8,8 @@
 
 def change_image_editor_mode(choice, cropped_image, resize_mode, width, height):
     if choice == "Mask":
-        return [gr.update(visible=False), gr.update(visible=True), gr.update(visible=False), gr.update(visible=True), gr.update(visible=False), gr.update(visible=False), gr.update(visible=True), gr.update(visible=True)]
-    return [gr.update(visible=True), gr.update(visible=False), gr.update(visible=True), gr.update(visible=False), gr.update(visible=True), gr.update(visible=True), gr.update(visible=False), gr.update(visible=False)]
+        return [gr.update(visible=False), gr.update(visible=True), gr.update(visible=False), gr.update(visible=True), gr.update(visible=False), gr.update(visible=True), gr.update(visible=True)]
```
Comment (anon-hlhl) [inline frontend/ui_functions.py:11]: Fixes https://github.com/hlky/stable-diffusion/issues/120

**Comment by hlky** [comment]
I'll test this shortly, thanks. Would you mind posting a link to the init image you used so I can just try and reproduce a similar result?

**Comment by anon-hlhl** [comment]
Can't find the exact image, but this one if approximately the same: http://i.huffpost.com/gen/3862740/images/o-KIM-JONG-UN-WAVING-facebook.jpg

For that example, I used DDIM (which is in that other PR, I personally believe the DDIM implemenation of masking is better as I understand it a bit more). As the can is a significantly different colour from his original hand, using a high Denoising Strength is necessary - something from 0.95 to 1

File: webui.py [hunk 2]
```diff
@@ -868,7 +868,28 @@ def process_images(
...
+                            try_loading_RealESRGAN(realesrgan_model_name)
+                        output, img_mode = RealESRGAN.enhance(np.array(init_img, dtype=np.uint8))
+                        init_img = Image.fromarray(output)
+                        init_img = init_img.convert('RGB')
+
+                        output, img_mode = RealESRGAN.enhance(np.array(init_mask, dtype=np.uint8))
+                        init_mask = Image.fromarray(output)
+                        init_mask = init_mask.convert('L')
+
+                    image = Image.composite(init_img, image, init_mask)
```
[the remark this answers is not in the document]
Comment (hlky) [reply webui.py:890]: Do you want to remove them and commit the change?

**Comment by hlky** [comment]
Works for me, good work!

**Comment by eva9900** [comment]
amazing

**Comment by hlky** [comment]
Will be merged to main repo soon @eva9900 just testing another PR that refactors some UI stuff

**Pull request merged**